README.md
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/tangle-frost/iota-qr-data/master/LICENSE) [![Build Status](https://travis-ci.org/tangle-frost/iota-qr-data.svg?branch=master)](https://travis-ci.org/tangle-frost/iota-qr-data) 

# Tangle Frost - IOTA QR Code Data Library

Library for helping convert IOTA data into QR codes.

# Installation

```shell
npm install @tangle-frost/iota-qr-data --save
```

# Documentation

Documentation for the code can be found in [docs](https://github.com/tangle-frost/iota-qr-data/blob/master/docs/README.md) folder.

# Contributing

Contributions are always welcome to the project. Feel free to raise issues, create pull requests or just make suggestions.

# Authors

Come and find us on the IOTA [Discord](https://discordapp.com/invite/fNGZXvh) development channels

* **Martyn Janes** - *obany* - ([https://github.com/obany](https://github.com/obany))

# Donations

IOTA donations are always welcome :smile:

![tangle-frost@tngl-me](https://cdn.tngl.me/tngl-me/tangle-frost/qr.svg)

# License

This project is licensed under the MIT License - see the [LICENSE](https://github.com/tangle-frost/iota-qr-data/blob/master/LICENSE) file for details.


CHANGELOG.md
# Changelog

## v0.0.9

* Version Bump

## v0.0.7

* Updated readme

## v0.0.6

* Widened range of targeted browser support

## v0.0.5

* Initial public release


test/data/svg.spec.ts
/**
 * Tests for Svg.
 */
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import * as chai from "chai";
import fs from "fs";
import { Svg } from "../../src/renderTypes/svg";

describe("Svg", () => {
    it("can be created", () => {
        const obj = new Svg();
        chai.should().exist(obj);
    });

    describe("render", () => {
        it("can render data", async () => {
            const qr = new QR();
            qr.addData("This is a test");
            const data = qr.generate();

            const svg = await Svg.render(data);

            const expected = fs.readFileSync("./test/renderTypes/results/svg1.svg");

            chai.expect(svg).to.equal(expected.toString());
        });
    });
});


test/data/txt.spec.ts
/**
 * Tests for Txt.
 */
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import * as chai from "chai";
import fs from "fs";
import { Txt } from "../../src/renderTypes/txt";

describe("Txt", () => {
    it("can be created", () => {
        const obj = new Txt();
        chai.should().exist(obj);
    });

    describe("render", () => {
        it("can render data", async () => {
            const qr = new QR();
            qr.addData("This is a test");
            const data = qr.generate();

            const txt = await Txt.render(data);

            const expected = fs.readFileSync("./test/renderTypes/results/txt1.txt");

            chai.expect(txt).to.equal(expected.toString());
        });
    });
});


test/data/bitmap.spec.ts
/**
 * Tests for Txt.
 */
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import * as chai from "chai";
import fs from "fs";
import { Bitmap } from "../../src/renderTypes/bitmap";

describe("Bitmap", () => {
    it("can be created", () => {
        const obj = new Bitmap();
        chai.should().exist(obj);
    });

    describe("render", () => {
        it("can render data as jpeg", async () => {
            const qr = new QR();
            qr.addData("This is a test");
            const data = qr.generate();

            const buffer = await Bitmap.render(data, "jpeg", 5, 10);

            const expected = fs.readFileSync("./test/renderTypes/results/jpeg1.jpg");

            chai.expect(buffer.toString()).to.equal(expected.toString());
        });

        it("can render data as bmp", async () => {
            const qr = new QR();
            qr.addData("This is a test");
            const data = qr.generate();

            const buffer = await Bitmap.render(data, "bmp", 5, 10);

            const expected = fs.readFileSync("./test/renderTypes/results/bmp1.bmp");

            chai.expect(buffer.toString()).to.equal(expected.toString());
        });

        it("can render data as png", async () => {
            const qr = new QR();
            qr.addData("This is a test");
            const data = qr.generate();

            const buffer = await Bitmap.render(data, "png", 5, 10);

            const expected = fs.readFileSync("./test/renderTypes/results/png1.png");

            chai.expect(buffer.toString()).to.equal(expected.toString());
        });
    });
});


webpack.config.js
const path = require('path');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
const packageJson = require('./package.json')

const isProd = process.env.NODE_ENV === 'production';

const plugins = [];
if (isProd) {
    plugins.push(new UglifyJsPlugin());
}

const calcPackageName = (packageJsonName) => packageJsonName.replace("@", "").replace("/", "-");
const calcRootName = (pkgName) => pkgName.replace(/-/g, " ").replace(/\b\w/g, l => l.toUpperCase()).replace(/ /g, "");

const pkgName = calcPackageName(packageJson.name);
const rootName = calcRootName(pkgName);

const orgName = packageJson.name.indexOf("@") === 0 ? packageJson.name.split("/")[0] : undefined;
const externals = [];

if (orgName) {
    externals.push(new RegExp(`^(${orgName})`));
}

module.exports = {
    entry: path.resolve(__dirname, './dist/index.js'),
    output: {
        path: path.resolve(__dirname, './pkg'),
        filename: pkgName + (isProd ? '.min' : '') + '.js',
        libraryTarget: 'umd',
        libraryExport: undefined,
        library: {
            root: rootName,
            amd: packageJson.name,
            commonjs: packageJson.name
        },
        umdNamedDefine: true,
        globalObject: 'typeof self !== \'undefined\' ? self : this'
    },
    target: "web",
    externals,
    mode: isProd ? "production" : "development",
    devtool: isProd ? undefined : "inline-source-map",
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: "babel-loader",
                    options: {
                        "presets": [
                            [
                                "@babel/preset-env", {
                                    "targets": {
                                        "browsers": [
                                            "Chrome >= 52",
                                            "FireFox >= 44",
                                            "Safari >= 7",
                                            "Explorer 11",
                                            "last 4 Edge versions"
                                        ]
                                    } 
                                }
                            ]
                        ],
                        "plugins": [
                            [
                                "@babel/plugin-transform-runtime",
                                {
                                    "helpers": false,
                                    "regenerator": true
                                }
                            ],
                            [
                                "babel-plugin-transform-builtin-extend",
                                {
                                    "globals": ["Error"]
                                }
                            ]
                        ]
                    }
                }
            }
        ]
    },
    node: {
        fs: "empty",
        path: "empty",
        console: false,
        global: false,
        process: false,
        __filename: false,
        __dirname: false,
        Buffer: false,
        setImmediate: false
    },
    plugins
};


src/data/addressQR.ts
import { ObjectHelper } from "@tangle-frost/iota-core/dist/helpers/objectHelper";
import { TrytesHelper } from "@tangle-frost/iota-core/dist/helpers/trytesHelper";
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import { QRRendererFactory } from "@tangle-frost/iota-qr-render/dist/factories/qrRendererFactory";

/**
 * Class to helper render addresses as QR.
 */
export class AddressQR {
    /**
     * Convert address data into a QR code raw data.
     * @param address The address to convert.
     * @param rendererType The type of render to use.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async renderRaw(
        address: string,
        rendererType: string,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<string | Uint8Array> {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(5);
        qr.addText(address);
        const qrCellData = qr.generate();

        return renderer.renderRaw(qrCellData, cellSize, marginSize);
    }

    /**
     * Convert address data into a QR code html element.
     * @param address The address to convert.
     * @param rendererType The type of render to use.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async renderHtml(
        address: string,
        rendererType: string,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<Element> {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(5);
        qr.addText(address);
        const qrCellData = qr.generate();

        return renderer.renderHtml(qrCellData, cellSize, marginSize);
    }
}


src/data/renderType.ts
/**
 * Class to helper render data for rinity as QR.
 */
export class Trinity {
    /**
     * Render the QR code data for tinity.
     * @param address The address trytes.
     * @param amount The amount for the transaction.
     * @param tag The tag for the transaction in trytes.
     * @param message The message for the transaction in trytes.
     * @returns The data for the rendered qr code.
     */
    public static transactionQR(address: string, amount?: number, tag?: string, message?: string): Promise<any> {
        if (address !== null && address !== undefined && /^[9A-Z]{81}$/.test(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (amount !== undefined || !Number.isInteger(amount)) {
            throw new Error("The amount must be an integer");
        }

        if (tag !== undefined && tag !== null) {
            if (tag.length > 27) {
                throw new Error(`The tag is too long, it should be at most 27 trytes, it is ${tag.length}`);
            }
            if (/^[9A-Z]+$/.test(tag)) {
                throw new Error("The tag does not appear to be in valid trytes format");
            }
        }

        const finalData: {
            address?: string;
            amount?: number;
            tag?: string;
            message?: string;
        } = {
            address,
            amount,
            message,
            tag
        };

        return undefined;
    }
}


src/data/trinityPaymentQR.ts
import { NumberHelper } from "@tangle-frost/iota-core/dist/helpers/numberHelper";
import { ObjectHelper } from "@tangle-frost/iota-core/dist/helpers/objectHelper";
import { StringHelper } from "@tangle-frost/iota-core/dist/helpers/stringHelper";
import { TrytesHelper } from "@tangle-frost/iota-core/dist/helpers/trytesHelper";
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import { QRRendererFactory } from "@tangle-frost/iota-qr-render/dist/factories/qrRendererFactory";
import { ITrinityPayment } from "../models/ITrinityPayment";

/**
 * Class to helper render data for trinity as QR.
 */
export class TrinityPaymentQR {
    /**
     * Create the QR code data for trinity payment data.
     * @param address The address trytes.
     * @param amountIota The amount for the transaction.
     * @param tagTrytes The tag for the transaction in trytes.
     * @param message The message for the transaction in plain text.
     * @returns The data for the trinity payment.
     */
    public static generatePaymentData(address: string, amountIota?: number, tagTrytes?: string, message?: string): ITrinityPayment {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        if (!ObjectHelper.isEmpty(amountIota)) {
            if (!NumberHelper.isInteger(amountIota)) {
                throw new Error("The amount must be an integer");
            }
        }

        if (!StringHelper.isEmpty(tagTrytes)) {
            if (!TrytesHelper.isTrytes(tagTrytes)) {
                throw new Error("The tag does not appear to be in valid trytes format");
            }
            if (tagTrytes.length > 27) {
                throw new Error(`The tag is too long, it should be at most 27 trytes, it is ${tagTrytes.length}`);
            }
        }

        let messageAscii;
        if (!StringHelper.isEmpty(message)) {
            messageAscii = StringHelper.encodeNonASCII(message);
        }

        return {
            address,
            amount: amountIota,
            message: messageAscii,
            tag: tagTrytes
        };
    }

    /**
     * Convert trinity payment data into a QR code raw data.
     * @param paymentData The payment data to convert.
     * @param rendererType The type of render to use.
     * @param qrTypeNumber The type number for qr code, controls the amount of data the QR can store.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async renderRaw(
        paymentData: ITrinityPayment,
        rendererType: string,
        qrTypeNumber: number = 16,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<string | Uint8Array> {
        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(qrTypeNumber);
        qr.addText(JSON.stringify(paymentData));
        const qrCellData = qr.generate();

        return renderer.renderRaw(qrCellData, cellSize, marginSize);
    }

    /**
     * Convert trinity payment data into a QR code html element.
     * @param paymentData The payment data to convert.
     * @param rendererType The type of render to use.
     * @param qrTypeNumber The type number for qr code, controls the amount of data the QR can store.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async renderHtml(
        paymentData: ITrinityPayment,
        rendererType: string,
        qrTypeNumber: number = 16,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<Element> {
        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(qrTypeNumber);
        qr.addText(JSON.stringify(paymentData));
        const qrCellData = qr.generate();

        return renderer.renderHtml(qrCellData, cellSize, marginSize);
    }
}


src/data/trinity.ts
import { NumberHelper } from "@tangle-frost/iota-core/dist/helpers/numberHelper";
import { ObjectHelper } from "@tangle-frost/iota-core/dist/helpers/objectHelper";
import { StringHelper } from "@tangle-frost/iota-core/dist/helpers/stringHelper";
import { TrytesHelper } from "@tangle-frost/iota-core/dist/helpers/trytesHelper";
import { QR } from "@tangle-frost/iota-qr-core/dist/qr";
import { QRRendererFactory } from "@tangle-frost/iota-qr-render/dist/factories/qrRendererFactory";
import { ITrinityPayment } from "../models/ITrinityPayment";

/**
 * Class to helper render data for trinity as QR.
 */
export class Trinity {
    /**
     * Create the QR code data for trinity.
     * @param address The address trytes.
     * @param amountIota The amount for the transaction.
     * @param tagTrytes The tag for the transaction in trytes.
     * @param message The message for the transaction in plain text.
     * @returns The data for the trinity payment.
     */
    public static generatePaymentData(address: string, amountIota?: number, tagTrytes?: string, message?: string): ITrinityPayment {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        if (!ObjectHelper.isEmpty(amountIota)) {
            if (!NumberHelper.isInteger(amountIota)) {
                throw new Error("The amount must be an integer");
            }
        }

        if (!StringHelper.isEmpty(tagTrytes)) {
            if (!TrytesHelper.isTrytes(tagTrytes)) {
                throw new Error("The tag does not appear to be in valid trytes format");
            }
            if (tagTrytes.length > 27) {
                throw new Error(`The tag is too long, it should be at most 27 trytes, it is ${tagTrytes.length}`);
            }
        }

        let messageAscii;
        if (!StringHelper.isEmpty(message)) {
            messageAscii = StringHelper.encodeNonASCII(message);
        }

        return {
            address,
            amount: amountIota,
            message: messageAscii,
            tag: tagTrytes
        };
    }

    /**
     * Convert trinity payment data into a QR code raw data.
     * @param paymentData The payment data to convert.
     * @param rendererType The type of render to use.
     * @param qrTypeNumber The type number for qr code, controls the amount of data the QR can store.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async paymentQRRaw(
        paymentData: ITrinityPayment,
        rendererType: string,
        qrTypeNumber: number = 16,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<string | Uint8Array> {
        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(qrTypeNumber);
        qr.addText(JSON.stringify(paymentData));
        const qrCellData = qr.generate();

        return renderer.renderRaw(qrCellData, cellSize, marginSize);
    }

    /**
     * Convert trinity payment data into a QR code html element.
     * @param paymentData The payment data to convert.
     * @param rendererType The type of render to use.
     * @param qrTypeNumber The type number for qr code, controls the amount of data the QR can store.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async paymentQRHtml(
        paymentData: ITrinityPayment,
        rendererType: string,
        qrTypeNumber: number = 16,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<Element> {
        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(qrTypeNumber);
        qr.addText(JSON.stringify(paymentData));
        const qrCellData = qr.generate();

        return renderer.renderHtml(qrCellData, cellSize, marginSize);
    }

    /**
     * Convert address data into a QR code raw data.
     * @param address The address to convert.
     * @param rendererType The type of render to use.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async addressQRRaw(
        address: string,
        rendererType: string,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<string | Uint8Array> {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(5);
        qr.addText(address);
        const qrCellData = qr.generate();

        return renderer.renderRaw(qrCellData, cellSize, marginSize);
    }

    /**
     * Convert address data into a QR code html element.
     * @param address The address to convert.
     * @param rendererType The type of render to use.
     * @param cellSize The size in pixels of each cell.
     * @param marginSize The margin size in pixels to leave around the qr code.
     * @param rendererOptions Any options you want to pass to the renderer.
     * @returns The render of the QR code in the requested format.
     */
    public static async addressQRHtml(
        address: string,
        rendererType: string,
        cellSize?: number,
        marginSize?: number,
        rendererOptions?: any): Promise<Element> {
        if (!TrytesHelper.isTrytes(address)) {
            throw new Error("The address does not appear to be in valid trytes format");
        }

        if (address.length !== 90) {
            throw new Error(`The address must be 90 trytes long and include the checksum, it is ${address.length}`);
        }

        const renderer = QRRendererFactory.instance().create(rendererType, rendererOptions);

        if (ObjectHelper.isEmpty(renderer)) {
            throw new Error(`The QR Renderer '${rendererType} is not available`);
        }

        const qr = new QR(5);
        qr.addText(address);
        const qrCellData = qr.generate();

        return renderer.renderHtml(qrCellData, cellSize, marginSize);
    }
}
